client/app/bundles/course/level/pages/LevelsIndex/LevelsManager.tsx

```diff
@@ -0,0 +1,150 @@
+import { useState } from 'react';
+import { defineMessages, FormattedMessage } from 'react-intl';
+import { Button, Slide, Typography } from '@mui/material';
+
+import { defaultComponentTitles } from 'course/translations.intl';
+import Page from 'lib/components/core/layouts/Page';
+import { useAppDispatch, useAppSelector } from 'lib/hooks/store';
+import useTranslation from 'lib/hooks/useTranslation';
+
+import LevelsTable from '../../components/LevelsTable';
+import { saveLevels } from '../../operations';
+import { getLocalState } from '../../selector';
+
+const translations = defineMessages({
+  unsavedChanges: {
+    id: 'course.level.Level.unsavedChanges',
+    defaultMessage: 'You have unsaved changes',
+  },
+  saveChanges: {
+    id: 'course.level.Level.saveChanges',
+    defaultMessage: 'Save',
+  },
+  reset: {
+    id: 'course.level.Level.reset',
+    defaultMessage: 'Reset',
+  },
+  saveSuccess: {
+    id: 'course.level.Level.saveSuccess',
+    defaultMessage: 'Levels Saved',
+  },
+  saveFailure: {
+    id: 'course.level.Level.saveFailure',
+    defaultMessage: 'Failed to save levels',
+  },
+});
+
+const LevelsManager = (): JSX.Element => {
+  const { canManage, levels } = useAppSelector(getLocalState);
+  const { t } = useTranslation();
+  const dispatch = useAppDispatch();
+
+  // Add 'deleted' property to each level in initial state
+  const [newLevels, setNewLevels] = useState(
+    levels.map((level) => ({ level, deleted: false })),
+  );
+  const [isSaving, setIsSaving] = useState(false);
+  const [isDirty, setIsDirty] = useState(false);
+
+  const handleAddLevel = (): void => {
+    setIsDirty(true);
+
+    const lastLevel = newLevels[newLevels.length - 1].level;
+    const newLevelThreshold = lastLevel ? lastLevel * 2 : 100;
+
+    const updatedLevels = [
+      ...newLevels,
+      { level: newLevelThreshold, deleted: false },
+    ];
+    setNewLevels(updatedLevels);
+  };
+
+  const handleThresholdChange = (index: number, newThreshold: string): void => {
+    const updatedLevels = [...newLevels];
+    updatedLevels[index + 1].level = parseInt(newThreshold, 10) || 0;
```

Comment:
any reason why the way you implement `updatedLevels` is different in here than inside the function `handleAddLevel`.

Just my personal opinion, please make it consistent, at least across this file

Reply:
Done

Reply:
Regarding this matter, please check carefully your implementation for `handleThresholdChange` and `handleDeleteLevels`. I tried just now to change the level and also to delete the level, then I suddenly got the error of `Cannot assign to read only property 'experiencePointsThreshold' of object '#<Object>'`

Please inspect this error

Reply:
Fixed